File: server/tackle_audio_test.go
package main

import (
	"reflect"
	"testing"
)

func TestTackleAudio(t *testing.T) {
	for team := 0; team < 2; team++ {
		for _, success := range []bool{false, true} {
			for _, carrying := range []bool{false, true} {
				s := initial()
				i, j := team*9+7, (1-team)*9+7
				for n := range s.Players {
					s.Players[n].Stun = 100
					s.Players[n].X, s.Players[n].Z, s.Players[n].aiWait = -20, -10, 100
				}
				p, q := &s.Players[i], &s.Players[j]
				p.X, p.Z, p.Stun, p.Action, p.ActionTime, p.FX, p.FZ = 0, 0, 0, 7, .16, 1, 0
				q.X, q.Z, q.Stun, q.FX, q.FZ = .5, 0, 0, 1, 0
				s.Ball.Owner, s.Ball.X, s.Ball.Z = -1, 10, 10
				if carrying {
					s.Ball.Owner = j
				}
				s.RNG = [2]uint32{65535, 65535}
				if success {
					s.RNG = [2]uint32{}
				}
				s.simulate(.04, [2]Input{}, [2]bool{true, true})
				expected := []int{29}
				if success {
					expected = append(expected, 4)
					if carrying {
						expected = append(expected, 24+team)
					}
				}
				actual := []int{}
				for _, e := range s.Events[:s.EventCount] {
					actual = append(actual, e.Kind)
				}
				if !reflect.DeepEqual(actual, expected) {
					t.Fatal(team, success, carrying, actual, expected)
				}
				id := s.Event.ID
				s.simulate(.04, [2]Input{}, [2]bool{true, true})
				if s.Event.ID != id {
					t.Fatal("resolved contact repeated")
				}
			}
		}
	}
}

File: src/tackle-audio.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, step } from "./game.js";
import { ArenaAudio } from "./audio.js";

test("tackle contact sounds on failure and possession signals only on a steal", () => {
  for (const team of [0, 1]) for (const success of [false, true]) for (const carrying of [false, true]) {
    const s = initial(), i = team * 9 + 7, j = (1 - team) * 9 + 7;
    for (const p of s.players) p.stun = 100;
    Object.assign(s.players[i], { x: 0, z: 0, stun: 0, action: 7, actionTime: .16, fx: 1, fz: 0 });
    Object.assign(s.players[j], { x: .5, z: 0, stun: 0, fx: 1, fz: 0 });
    Object.assign(s.ball, { owner: carrying ? j : -1, x: 10, z: 10 });
    s.rng = success ? [0, 0] : [65535, 65535];
    step(s, .04, {}, [true, true], {});
    const expected = success ? carrying ? [29, 4, team === 0 ? 24 : 25] : [29, 4] : [29];
    assert.deepEqual(s.events.map(e => e.kind), expected);
    const audio = new ArenaAudio(), heard = [];
    audio.play = kind => heard.push(kind);
    audio.observe(s, true);
    audio.observe(s, true);
    assert.deepEqual(heard.filter(k => k !== "kickoff"), expected);
    const id = s.event.id;
    step(s, .04, {}, [true, true], {});
    assert.equal(s.event.id, id, "resolved contact must not replay each tick");
  }
});
